consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,637 @@
+	PrevoteListener   Listener[Prevote[H, A], V, H, A]
+	PrecommitListener Listener[Precommit[H, A], V, H, A]
+}
+
+type Broadcasters[V Hashable[H], H Hash, A Addr] struct {
+	ProposalBroadcaster  Broadcaster[Proposal[V, H, A], V, H, A]
+	PrevoteBroadcaster   Broadcaster[Prevote[H, A], V, H, A]
+	PrecommitBroadcaster Broadcaster[Precommit[H, A], V, H, A]
+}
+
+type Tendermint[V Hashable[H], H Hash, A Addr] struct {
+	nodeAddr A
+
+	state state[V, H] // Todo: Does state need to be protected?
+	// Todo: maintain a separate message set for future round messages.
+	// Which are replayed when moving to the next round.
+	messages messages[V, H, A]
+
+	timeoutPropose   timeoutFn
+	timeoutPrevote   timeoutFn
+	timeoutPrecommit timeoutFn
+
+	application Application[V, H]
+	blockchain  Blockchain[V, H, A]
+	validators  Validators[A]
+
+	listeners    Listeners[V, H, A]
+	broadcasters Broadcasters[V, H, A]
+
+	scheduledTms []timeout
+	timeoutsCh   chan timeout
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+type state[V Hashable[H], H Hash] struct {
+	height uint
+	round  uint
+	step   step
+
+	lockedValue *V
+	validValue  *V
+
+	// The default value of lockedRound and validRound is -1. However, using int for one value is not good use of space,
+	// therefore, uint is used and nil would represent -1.
+	lockedRound *uint
+	validRound  *uint
+
+	// The following are round level variable therefore when a round changes they must be reset.
+	line34Executed bool
+	line36Executed bool
+	line47Executed bool
+}
+
+// Todo: Add Slashers later
+func New[V Hashable[H], H Hash, A Addr](addr A, app Application[V, H], chain Blockchain[V, H, A], vals Validators[A],
+	listeners Listeners[V, H, A], broadcasters Broadcasters[V, H, A], tmPropose, tmPrevote, tmPrecommit timeoutFn,
+) *Tendermint[V, H, A] {
+	return &Tendermint[V, H, A]{
+		nodeAddr:         addr,
+		state:            state[V, H]{height: chain.Height()},
+		messages:         newMessages[V, H, A](),
+		timeoutPropose:   tmPropose,
+		timeoutPrevote:   tmPrevote,
+		timeoutPrecommit: tmPrecommit,
+		application:      app,
+		blockchain:       chain,
+		validators:       vals,
+		listeners:        listeners,
+		broadcasters:     broadcasters,
+		scheduledTms:     make([]timeout, 0),
+		timeoutsCh:       make(chan timeout),
+		quit:             make(chan struct{}),
+	}
+}
+
+//nolint:gocyclo,funlen
+func (t *Tendermint[V, H, A]) Start() {
+	t.wg.Add(1)
+	go func() {
+		defer t.wg.Done()
+
+		t.startRound(0)
+
+		// Todo: check message signature everytime a message is received.
+		// For the time being it can be assumed the signature is correct.
+		i := 0
+		for {
+			fmt.Println("iteration: ", i)
+			select {
+			case <-t.quit:
+				fmt.Println("quit", t.state.step)
+				return
+			case p := <-t.listeners.ProposalListener.Listen():
+				if p.Height < t.state.height {
+					continue
+				}
+
+				if p.Height > t.state.height { //nolint:staticcheck
+					continue
+				}
+
+				if p.Round > t.state.round {
```

Comment:
Can the proposers round also be behind ours? Like with the height above?

Reply:
Technically, yes but it is an extremely unlikely scenario.

 At any moment in time, you need 2/3 of the network to make progress. If a node falls behind it means that node is not required to form a quorum. If this does happen, there are mechanisms in place to help the proposer catch up if that takes too long then timeout will eventually change the round which changes the proposer thus allowing the network to make progress.